Enable KubernetesExecutor, update to Airflow 1.10.11, and further fixes

noseka1 opened the pull request:
Many improvements for the docker-airflow container image. Please, refer to the individual commits for details.

noseka1 committed:
Fix /entrypoint.sh: line 41: AIRFLOW__CORE__LOAD_EXAMPLES: command not found

noseka1 committed:
Allow Celery worker to run on OpenShift

Celery worker refuses to run when executed as gid=0 egid=0 on OpenShift

$ exec airflow worker
Running a worker with superuser privileges when the
worker accepts messages serialized with pickle is a very bad idea!

If you really want to continue then you have to set the C_FORCE_ROOT
environment variable (but please think about this before you do).

User information: uid=1000660000 euid=1000660000 gid=0 egid=0

noseka1 committed:
Update to Airflow 1.10.11

The KubernetesExecutor in this version is much improved. There was
an issue with the KubernetesExecutor in version 1.10.10. It was
setting the worker pod to run with uid=0 by default. It was not
possible to prevent KubernetesExecutor from setting the uid.
As a result, worker pods would fail to start on OpenShift when
using the default restricted SCC.

Airflow 1.10.11 correctly refrains from setting any uid for the pods,
unless the user explicitely requests it.

The issue seen when using KubernetesExecutor in version 1.10.10
on OpenShift while using restricted SCC:

pk-cluster-scheduler-0 scheduler [2020-07-12 23:09:34,297] {{pod_launcher.py:84}} ERROR - Exception when attempting to create Namespaced Pod.
pk-cluster-scheduler-0 scheduler Traceback (most recent call last):
pk-cluster-scheduler-0 scheduler   File "/usr/local/lib/python3.7/site-packages/airflow/contrib/kubernetes/pod_launcher.py", line 81, in run_pod_async
pk-cluster-scheduler-0 scheduler     resp = self._client.create_namespaced_pod(body=req, namespace=pod.namespace, **kwargs)
pk-cluster-scheduler-0 scheduler   File "/usr/local/lib/python3.7/site-packages/kubernetes/client/api/core_v1_api.py", line 6174, in create_namespaced_pod
pk-cluster-scheduler-0 scheduler     (data) = self.create_namespaced_pod_with_http_info(namespace, body, **kwargs)  # noqa: E501
pk-cluster-scheduler-0 scheduler   File "/usr/local/lib/python3.7/site-packages/kubernetes/client/api/core_v1_api.py", line 6265, in create_namespaced_pod_with_http_info
pk-cluster-scheduler-0 scheduler     collection_formats=collection_formats)
pk-cluster-scheduler-0 scheduler   File "/usr/local/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 345, in call_api
pk-cluster-scheduler-0 scheduler     _preload_content, _request_timeout)
pk-cluster-scheduler-0 scheduler   File "/usr/local/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 176, in __call_api
pk-cluster-scheduler-0 scheduler     _request_timeout=_request_timeout)
pk-cluster-scheduler-0 scheduler   File "/usr/local/lib/python3.7/site-packages/kubernetes/client/api_client.py", line 388, in request
pk-cluster-scheduler-0 scheduler     body=body)
pk-cluster-scheduler-0 scheduler   File "/usr/local/lib/python3.7/site-packages/kubernetes/client/rest.py", line 278, in POST
pk-cluster-scheduler-0 scheduler     body=body)
pk-cluster-scheduler-0 scheduler   File "/usr/local/lib/python3.7/site-packages/kubernetes/client/rest.py", line 231, in request
pk-cluster-scheduler-0 scheduler     raise ApiException(http_resp=r)
pk-cluster-scheduler-0 scheduler kubernetes.client.rest.ApiException: (403)
pk-cluster-scheduler-0 scheduler Reason: Forbidden
pk-cluster-scheduler-0 scheduler HTTP response headers: HTTPHeaderDict({'Audit-Id': 'd0f97a01-3c49-45b0-8f6d-83577355e9bd', 'Content-Type': 'application/json', 'Date': 'Sun, 12 Jul 2020 23:09:34 GMT', 'Content-Length': '677'})
pk-cluster-scheduler-0 scheduler HTTP response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods \"examplebashoperatorrunme2-4fd9b84d6c67438e8e522ca573a4e10a\" is forbidden: unable to validate against any security context constraint: [fsGroup:
Invalid value: []int64{0}: 0 is not an allowed group spec.initContainers[0].securityContext.securityContext.runAsUser: Invalid value: 65533: must be in the ranges: [1000660000, 1000669999] spec.containers[0].securityContext.securityContext.runAsUser: Invalid value: 0: must be in the
ranges: [1000660000, 1000669999]]","reason":"Forbidden","details":{"name":"examplebashoperatorrunme2-4fd9b84d6c67438e8e522ca573a4e10a","kind":"pods"},"code":403}

noseka1 committed:
Always delete worker pods after they are complete

Avoid failed worker pods piling up in the etcd database. User can
enable this feature when desired.

noseka1 committed:
Don't load default connections in production

noseka1 committed:
KubernetesExecutor fails to create worker pods on OpenShift

The default restricted SCC doesn't allow setting uid, gid for the
pods.

noseka1 committed:
Use upgradedb instead of initdb to initialize the database

Airflow best practices recommend using upgradedb instead of initdb for
production databases:

Do not use airflow initdb as it can create a lot of default connection,
charts, etc. which are not required in production DB.

Link:
https://github.com/apache/airflow/blob/1.10.11/docs/best-practices.rst#database-backend

noseka1 committed:
Delay scheduler start instead of webserver start

Database was originally initialized by the scheduler container.
Later on, the database initizalization was moved to the webserver
container. This commit fixes up the change.

vpavlin commented on docker-compose-CeleryExecutor.yml:
```diff
@@ -16,7 +16,7 @@ services:
 
     webserver:
-        image: puckel/docker-airflow:1.10.9
+        image: puckel/docker-airflow:1.10.11
```
Does it make sense to change it here even though the image does not exist in dockerhub?

noseka1 replied:
Good point, it doesn't make sense. I can change it back to puckel/docker-airflow:1.10.9. Or probably better would be to point it to quay.io/opendatahub/docker-airflow:openshift?

vpavlin replied:
I think pointing to quay.io might be good. We can go with a specific version if we want, just need to create a new tag in the registry.

noseka1 replied:
If there is a 1.10.11 tag in quay.io I could also use that tag in the airflow-on-k8s-operator.

I am also thinking to switch to using the Red Hat's database images in docker-compose-CeleryExecutor.yml and docker-compose-LocalExecutor.yml.

So, the overall strategy would be to use the same images in both docker-airflow and airflow-on-k8s-operator. I think that would make sense.

vpavlin replied:
Sure, sounds good...though I am not sure where this tag appears in airflow-on-k8s-operator

noseka1 replied:
I meant in the airflow-on-k8s-operator/api/v1alpha1/airflowcluster_types.go to change:

```
  defaultFlowerImage      = "quay.io/opendatahub/docker-airflow"
  defaultFlowerVersion    = "openshift"
  defaultSchedulerImage   = "quay.io/opendatahub/docker-airflow"
  defaultSchedulerVersion = "openshift"
  defaultUIImage          = "quay.io/opendatahub/docker-airflow"
  defaultUIVersion        = "openshift"
  defaultWorkerImage      = "quay.io/opendatahub/docker-airflow"
  defaultWorkerVersion    = "openshift"
```
to

```
  defaultFlowerImage      = "quay.io/opendatahub/docker-airflow"
  defaultFlowerVersion    = "1.10.11"
  defaultSchedulerImage   = "quay.io/opendatahub/docker-airflow"
  defaultSchedulerVersion = "1.10.11"
  defaultUIImage          = "quay.io/opendatahub/docker-airflow"
  defaultUIVersion        = "1.10.11"
  defaultWorkerImage      = "quay.io/opendatahub/docker-airflow"
  defaultWorkerVersion    = "1.10.11"
```
Plus also change this commit to use "1.10.11" instead of the "openshift" tag:
https://github.com/noseka1/airflow-on-k8s-operator/commit/3e7cdbd09a42d9c0e06908e972351035d9d816a0

vpavlin replied:
Ah, ok, yeah, sounds good

noseka1 replied:
I updated the images in both the docker-airflow and the airflow-on-k8s-operator repos.

vpavlin commented:
I tested it by deploying via ODH and it seems to work fine - I was able to execute some simple workflows. So once we have the right images specified, I am happy to merge.

Awesome work, @noseka1!

noseka1 committed:
Use Airflow container images provided by OpenDataHub

Image puckel/docker-airflow:1.10.11 doesn't exist at this time.

noseka1 committed:
Docker-compose-LocalExecutor won't start Airflow properly

Tested the fix using the command:
podman-compose -f docker-compose-LocalExecutor.yml up

noseka1 commented:
@vpavlin Thanks! I updated the images in both the docker-airflow and the airflow-on-k8s-operator repos. Let me know if anything else needs to be done. This PR should be ready for merge now. Thanks.
